File: arrays_and_strings/189-rotate_array.cpp
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

class Solution {
public:
  void rotate(vector<int> &nums, int k) {
    int n = nums.size();
    k %= n;
    int cycles = gcd(k, n);
    for (int i = 0; i < cycles; i++) {
      int currIdx = i;
      int currEle = nums[currIdx];
      while (true) {
        int nextIdx = (currIdx + k) % n;
        int nextEle = nums[nextIdx];

        nums[nextIdx] = currEle;

        currEle = nextEle;
        currIdx = nextIdx;

        if (currIdx == i)
          break;
      }
    }

    for (int i = 0; i < n; i++) {
      cout << nums[i] << " ";
    }
  }
};

int main() {
  vector<int> nums = {1, 2, 3, 4, 5, 6, 7, 8};
  Solution().rotate(nums, 3);
  return 0;
}

File: arrays_and_strings/189-rotate_array.py
from math import gcd
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k %= n
        cycles = gcd(n, k)

        for i in range(cycles):
            currIdx = i
            currElem = nums[currIdx]

            while True:
                nextIdx = (currIdx + k) % n
                nextElem = nums[nextIdx]

                nums[nextIdx] = currElem

                currElem = nextElem

                currIdx = nextIdx

                if currIdx == i:
                    break


nums: List[int] = [1, 2, 3, 4, 5, 6, 7]
Solution().rotate(nums, 3)
print(nums)
